ICS33: Multi-hop Channel Spec (#882)
cosmos/ibc: spec/core/ics-033-multi-hop/README.md

2. Iterate through connection proof data and verify each connectionEnd is in the OPEN state. Check connectionHops during channel handshake.
3. Starting with known `ConsensusState[N-1]` at the given `proofHeight` on `ChainN` prove the prior chain's consensus and connection state.
4. Repeat step 3, proving `ConsensusState[i-1]` and `Conn[i-1,i]` until `ConsensusState0` on the source chain is proven.
   - Start with i = N-1
   - ConsensusState <= ConsensusState[i-1] on Chain[i]
   - ConsensusProofs[i].Proof.VerifyMembership(ConsensusState.GetRoot(), ConsensusProofs[i].Key, ConsensusProofs[i].Value)
   - ConnectionProofs[i].Proof.VerifyMembership(ConsensusState.GetRoot(), ConsensusProofs[i].Key, ConsensusProofs[i].Value)
   - Set ConsensusState from unmarshalled ConsensusProofs[i].Value
5. Finally, prove the expected channel or packet commitment in `ConsensusState0`

For more details see [ICS4](https://github.com/cosmos/ibc/tree/main/spec/core/ics-004-channel-and-packet-semantics).

### Multi-hop Proof Verification Pseudo Code

Pseudocode proof verification of a channel between chains `A -> B -> C` .

```typescript

> function verifyMembership(

Reviewer: I think we need more explicit pseudocode here. Similar to what you have above in proof generation, where you loop through the consensus state and connection state proofs and verify them

    root: CommitmentRoot, 

>     commitments: map[CommmitmentPath]bytes,

Reviewer: Perhaps this should be inside the MultiHopProof?

Reviewer: I guess if already does by virtue of having the data in `ProofData`

    batchProof: CommitmentProof) {
    // verifies the commitments in the commitment root using a batch proof
}

// Prove B's connectionState and consensusState on C
commitmentsB = map[CommitmentPath]bytes{
    connectionB: bytes,
    consensusStateB: bytes,
}
verifyMembership(
    rootC: CommitmentRoot, 
    commitmentsB: map[CommitmentPath]bytes, 
    batchProofB: CommitmentProof)

// Verify that the clientID from the connectionState is the same as in the consenstateState commitment path
// Prove A's connectionState and consensusState on B
commitmentsA = map[CommitmentPath]bytes{
    connectionA: bytes,
    consensusStateA: bytes,
}
verifyMembership(
    rootB: CommitmentRoot, 
    commitmentsA: map[CommitmentPath]bytes, 
    batchProofA: CommitmentProof)

// Repeat for all intermediate roots

// Verify the proven connection ID's match the connections in connectionHops

> verifyConnectionIDs(connectionHops: []string, provenConnectionIDs: []string)

Reviewer: Where is `provenConnectionIDs` coming from?

        abortTransactionUnless(parts[len(parts)-1] == connectionHops[i+1])

        // 2. check that the connectionEnd's are in the OPEN state.
        abortTransactionUnless(connectionEnd.GetState() == int32(connectiontypes.OPEN))
    }

    // create prefixed key for proof verification
    prefixedKey := abortTransactionUnless(commitmenttypes.ApplyPrefix(prefix, commitmenttypes.NewMerklePath(key)))
 
    // verify each consensus state and connection state starting going from Z --> A
    // finally verify the keyproof on A within B's verified view of A's consensus state.
    abortTransactionUnless(VerifyMultiHopProofMembership(consensusState, &proofs, &prefixedKey, value))
}

// VerifyMultiHopProofMembership verifies a multihop membership proof including all intermediate state proofs.
func VerifyMultiHopProofMembership(
    consensusState exported.ConsensusState,
    proofs *MsgMultihopProofs,
    prefixedKey commitmenttypes.MerklePath,
    value []byte,
) {
    abortTransactionUnless(len(proofs.ConsensusProofs) >= 1)
    abortTransactionUnless(len(proofs.ConnectionProofs) == len(proofs.ConsensusProofs))
 

>     abortTransactionUnless(VerifyMultiHopConsensusStateProof(consensusState, proofs.ConsensusProofs, proofs.ConnectionProofs))

Reviewer: Took me a while to understand this was correct.

There are two consensus states here.

`consensusState` is the consensus state stored on the destination chain to prove the first hop in the multiHop chain.

`consState` is the consensus state stored on the first chain in the multiHop link used to prove the final value.

These should be better named so its easier to understand the difference and where they are used

    keyProof := abortTransactionUnless(Unmarshal(proofs.KeyProof.Proof))
    
    // the consensus state of the source chain (chain[0] on chain[1])
    consState := abortTransactionUnless(UnmarshalInterface(proofs.ConsensusProofs[0].Value))

    abortTransactionUnless(keyProof.VerifyMembership(
        commitmenttypes.GetSDKSpecs(),
        consState.GetRoot(),
        prefixedKey,
        value,
    ))
}

// VerifyMultiHopConsensusStateProof verifies the consensus state of each consecutive consensus/connection state starting
// from chain[N-1] on the destination (chain[N]) and finally proving the source chain consensus/connection state.
func VerifyMultiHopConsensusStateProof(
 consensusState exported.ConsensusState,
 consensusProofs []*MultihopProof,
 connectionProofs []*MultihopProof,
) {
    // reverse iterate through proofs to prove from destination to source
    for i := len(consensusProofs) - 1; i >= 0; i-- {
        consStateProof := consensusProofs[i]
        connectionProof := connectionProofs[i]

>         conState := abortTransactionUnless(UnmarshalInterface(consStateProof.Value))

Reviewer: Move this down to line 334 for understandability

        // prove the consensus state of chain[i] in chain[i+1]
        consensusProof := abortTransactionUnless(Unmarshal(consStateProof.Proof))
        abortTransactionUnless(consensusProof.VerifyMembership(
            commitmenttypes.GetSDKSpecs(),
            consensusState.GetRoot(),
            *consStateProof.PrefixedKey,
            consStateProof.Value,
        ))

        // prove the connection state of chain[i] in chain[i+1]
        connectionProof := abortTransactionUnless(Unmarshal(connectionProof.Proof))
        abortTransactionUnless(connectionProof.VerifyMembership(
            commitmenttypes.GetSDKSpecs(),
            consensusState.GetRoot(),
            *connectionProof.PrefixedKey,

>             connectionProof.Value, // this should be from connectionHops

Reviewer 2: Not clear to me, do we check anywhere that the consensus states are from clients associated with the actual connections?

Reviewer 3: So, we check each connection is OPEN and matching the connections specified in the channel connectionHops field [here](https://github.com/polymerdao/ibc/blame/bo/ics-32/spec/core/ics-033-multi-hop/README.md#L276).

We also check that each of those provided connectionEnds are within the provided consensus states [here](https://github.com/polymerdao/ibc/blame/bo/ics-32/spec/core/ics-033-multi-hop/README.md#L311)

At a hight level we prove:
- The connectionHops used to establish the channel match the connections provided in the multihop proof and are OPEN
- the connectionEnds are proven within the provided consensus states
- the consensus states are proven recursively starting from the trusted consensus state on the receiving chain.

Does that suffice?

Reviewer 2: Two clients may have the same consensus state. So I was looking for an explicit check that `connectionEnd.ClientId` is the same as the `clientId` for which the consensus proof was collected, i.e. same as the ID in its key path. Check would be something like:
```
consensusProofClientId = strings.Split(consensusProof.PrefixedKey.GetKeyPath()[len(consensusProof.PrefixedKey.KeyPath)-1], "/")..
abortTransactionUnless(connectionEnd.ClientId == consensusProofClientId)
```

Reviewer 3: Yeah, I think we will likely need to bundle clientStates with the multihop proof as well per offline discussion. In this case we should definitely verify the clientid of the clientstate matches the consensus/connection state.

Reviewer 3: @ancazamfir also, FYI, I'm not sure if you saw this. Here is the offline doc summarizing the review comments: https://docs.google.com/document/d/1lFxGZwapWANirgCO3NthsDGHgmBdSo_71alj14rAhnU/edit#

Reviewer 3: Feel free to drop comments there and we can enshrine the decision in GH.

Reviewer 2: thanks, i will take a look and provide comments there

Reviewer 3: To summarize the offline discussion and google doc collaboration:

- delayPeriod will be enforced by computing the maximum delayPeriod across all connections in the channel path and enforcing that delayPeriod on the final consensusState on the verifier. This should suffice because by construction all recursively embedded consensus states must have existed for at least as long as the final consensus state.
- Client states are NOT required to be bundled with the multi-hop proof. Instead a new chanCloseFrozen() method will be introduced to allow relayers to post a multi-hop proof of a frozen client state starting from the misbehaving chain in the channel path. The proof is not symmetric, but relayers can send a unique proof (representing the hops from each end to the misbehaving chain) to actively freeze the channel.
- Frozen channels can not process packets and a new FROZEN state could be introduced to help distinguish this situation.
- If a client is frozen in the channel path then it would not be possible to send newly committed packets as the consensus state associated with the frozen client would not be updatable making it impossible to create new multi-hop proofs.

Hopefully I represented that correctly and didn't miss anything.

// Consensus and Connection proofs are ordered from source to destination but not including the
// chain[N-1] consensus state on chainN (dest) since it is already known on the destination chain.
type MultihopProof struct {
    KeyProofIndex uint32           // the index of the consensus state to prove the key/value on, normally 0
    KeyProof *ProofData            // the key/value proof on the source chain (consensusState[0])
    ConsensusProofs []*ProofData   // array of consensus proofs starting with proof of consensusState[0] state root on chain1 (consensusState[1])
    ConnectionProofs []*ProofData  // array of connection proofs starting with proof of conn[1,0] on chain1 (consensusState[1])
}

// GenerateMultihopProof generates proof of a key/value at the proofHeight on source chain (chain0). 
func GenerateMultihopProof(chains []*Chain, key string, value []byte, proofHeight exported.Height) *MultihopProof {

    abortTransactionUnless(len(chains) > 2)

    var multihopProof MultihopProof
    chain0 := chains[0] // source chain
    chain1 := chains[1] // first hop chain
 
    height01 := chain1.GetClientStateHeight(chain0) // height of chain0's client state on chain1

>     abortTransactionUnless(height01 >= proofHeight) // ensure that chain0's client state is update to date

Reviewer 3: Hmm, looking at this pseudo code I think it needs some clarification.

Really what this code is trying to do is ensure that each client in the multi-hop channel path is up to date with respect to the prior chain. The key to be proven on the source chain (chain0) is committed at some height so can be proven in chain0's consensusState on chain1 at that specific height. This is why the check is ">=". Does that make sense or maybe I missed something?

Its a bit confusing thought  since this is not really the `proofHeight` which has a different meaning. I think maybe it could be called `keyHeight`? Each subsequent client state needs to be up to date to ensure the prior chain's consensusState is included and that the consensusState at `keyHeight` on chain1 includes the key/value on the source chain.

The proofHeight during verification is used to indicate the height of the N-1'th consensus/connection on the N'th (receiving) chain.

    // query the key/value proof on the source chain at the proof height

>     keyProof, _ := chain0.QueryProofAtHeight([]byte(key), int64(proofHeight.GetRevisionHeight()))

Reviewer 3: It is part of the `ibctesting` framework (chain.go) though I'm not sure if it is defined in any ICS. Do you suggest explicitly defining it in this case?

>     keyProof, _ := chain0.QueryProofAtHeight([]byte(key), int64(proofHeight.GetRevisionHeight()))

Reviewer 2: Not really, I was curious is there is one. Maybe just declare the primitives in a prior section. This could also be simplified to pseudocode, smthg like `chain0.QueryProofAtHeight(key, proofHeight)` or `QueryProofAtHeight(chain0, key, proofHeight)`

        Proof: keyProof,
    }

    return &proof
}

// GenerateFrozenClientProof generate a multihop proof of a frozen channel given a set of chains starting from the misbehaving chain.
// Chains are provided in order from the misbehaving chain to the verifying chain.
func GenerateFrozenClientProof(chains []*Chain, proofHeight exported.Height) *MultihopProof {

    abortTransactionUnless(len(chains) > 2)

    // generate and assign consensus, clientState, and connection proofs
    var proof MultihopProof
    proof.ConsensusProofs = GenerateConsensusProofs(chains)
    proof.ConnectionProofs = GenerateConnectionProofs(chains)

    chain0 := chains[0] // misbehaving chain
    chain1 := chains[1] // next hop chain

    // check a consensus state at proof height is present on chain 1
    consensusStateAtProofHeight := chain1.GetConsensusStateAtHeight(chain0, proofHeight)
    abortTransactionUnless(consensusStateAtProofHeight != nil)

    // connectionEnd on misbehaving chain representing the counterparty connection on the next chain
    counterpartyConnectionEnd := abortTransactionUnless(Unmarshal(proof.ConnectionProofs[0].Value))

> 

Reviewer 2: At first look this doesn't seem to work imho. Could we have a section going through an example? I think it would be useful as the proof generation in general is complex and for this case in particular.
So again let's take this example with connection IDs showing. Assume the client used by `bc` is frozen

`A (a) --- (ba) B (bc) xx --- (cb) C (cd) --- (dc) D`

So we call:
1. `GenerateFrozenClientProof([B, A])` -> proofs will be sent to A
- In this case A will get connection proofs `[bc, a]` and will take the client of `bc`.

2. `GenerateFrozenClientProof([B, C, D])` -> proofs will be sent to D
- D will get proofs for `[cb, dc]` but should use the client of the counterparty of `cb`.

So 1. and 2. don't look the same wrt to proof collection logic. Same for verification in `chanCloseFrozen`.
Intuitively I was expecting some kind of indication of the "direction" being froze, not sure we can generalize for both 1 and 2. But I might be proven wrong.

Reviewer 3: Yes, your assessment is correct. We made a decision to only allow the proof to work on the side with the frozen client (proof logic to A in step 1 from your example). We came to the same conclusion that it would be hard to generalize though this is certainly open for consideration.

The consequence of this choice is that in your example above only the channel on chain A could be forcefully frozen. In order for the channel on chain D to be frozen `dc` would need to be frozen (although in this example it would not be necessary since chain D would already see the frozen client state).

So in order to freeze both ends of the channel, evidence must be submitted to freeze the client state on both sides of the misbehaving chain. This way the proof logic is always the same, and the tradeoff is that the channel can only be frozen on the side of the channel path containing the frozen client.

Thoughts?

Reviewer 2: _(A clarification that in my example those IDs were for connections but yes, we have similar for clientIDs)_ 

When we discussed this offline, I was asking the same question and feedback from @AdityaSripal was that we should close both ends. I agree with that, we should close both ends.
We cannot freeze a client on the other "side" as that client/ connection may be valid and used by other single or multi-hop channels.

I think we can make this function work if we add an explicit flag, e.g. `i` (incoming) to the last proof. By default proofs are `o`, i.e. outgoing.

_(Note that I flipped the proofs below as well, i.e. they can be verified from first to last)_
So for 2. the proof could be something like `[dc, cb, bc(i)]` vs for 1. `[a, bc]`
If the last proof is marked `i` then the client on the incoming/ counterparty connection end is frozen, i.e. `bc` must be for the counterparty of `cb`.

Reviewer 3: I have one question with this approach. If `ClientStateC` on `ChainB` (represented as `bc`) is frozen, would that be provable on `ConsensusStateB` on `ChainC`? Once the client is frozen there will not be any more consensus updates so would `ConsensusStateB` be frozen just prior to the freeze or would that state be included in the state?

That said, I went back and looked at the implementation for the ChanCloseFrozen proof, and I actually think the existing logic can support channel freezing on both ends of the channel with very small changes by using the new `KeyProofIndex` field of the MultihopProof struct to also indicate the directionality of the proof. Going back to your example above:

`A (a) --- (ba) B (bc) xx --- (cb) C (cd) --- (dc) D`

1. For A, its `multihopConnectionEnd` is `bc`. The ClientID for ClientBC should be `multihopConnectionEnd.Counterparty.ClientD` and we can use this logic to determine the client state key if the `KeyProofIndex != 0`

2. For D, its `multihopConnectionEnd` is `cb`. The ClientID for ClientBC should be `multihopConnectionEnd.ClientID` and we can use this to determine the key for the client state if `KeyProofIndex == 0`

`KeyProofIndex` is a new field introduced to keep the multi-hop verification logic the same while allowing verification of keys within intermediate consensus states in the proof chain.

Reviewer 2: > I have one question with this approach. If `ClientStateC` on `ChainB` (represented as `bc`) is frozen, would that be provable on `ConsensusStateB` on `ChainC`? Once the client is frozen there will not be any more consensus updates so would `ConsensusStateB` be frozen just prior to the freeze or would that state be included in the state?
> 

Not sure I understand. If the client for `bc` is frozen at height `hb` we can still update the client for `cb` on C with a consensus state of chain B at a height `hb', hb'>hb`. Then we can query B's client for `bc` with proof at height `hb'`.

> That said, I went back and looked at the implementation for the ChanCloseFrozen proof, and I actually think the existing logic can support channel freezing on both ends of the channel with very small changes by using the new `KeyProofIndex` field of the MultihopProof struct to also indicate the directionality of the proof. Going back to your example above:
> 
> `A (a) --- (ba) B (bc) xx --- (cb) C (cd) --- (dc) D`
> 
> 1. For A, its `multihopConnectionEnd` is `bc`. The ClientID for ClientBC should be `multihopConnectionEnd.Counterparty.ClientD` and we can use this logic to determine the client state key if the `KeyProofIndex != 0`

shouldn't be `multihopConnectionEnd.ClientID` in this case? and ``KeyProofIndex == 0`
> 2. For D, its `multihopConnectionEnd` is `cb`. The ClientID for ClientBC should be `multihopConnectionEnd.ClientID` and we can use this to determine the key for the client state if `KeyProofIndex == 0`
> 

shouldn't be `multihopConnectionEnd.Counterparty.ClientD` and `KeyProofIndex != 0`

> `KeyProofIndex` is a new field introduced to keep the multi-hop verification logic the same while allowing verification of keys within intermediate consensus states in the proof chain.

So iiuc, `KeyProofIndex` is non-zero in case 2 of proofs for closing channels due to on-path client being frozen.

Reviewer 2: I was thinking about this when reviewing and got stuck exactly here...
> Here, I would say that a relayer should first freeze the clients of C on the direct counterparties. Namely, (bc) and (dc) should be frozen.

hermes currently submits evidence only once a client is updated. If `bc` client is updated (by some other relayer) hermes compares the header used in the update with the header it gets from the C full node it connects to. Then it submits `MsgMisbehaviour` but to `bc` only.

The relayer could try to freeze all clients of C as you point out. But there is no guarantee that this works, in the example, the header that updated `bc` might not be verifiable on the `dc` client because the clients may have different security parameters (specifically the trusting level). So in theory we have a small probability channel on E will not be closed because we cannot freeze `dc`. 
 
Can we live with this in practice? It would simplify the spec quite a bit, we could eliminate `KeyProofIndex`, etc.

The spec for channel handshakes and packets remains the same. See [ICS 4](https://github.com/cosmos/ibc/tree/main/spec/core/ics-004-channel-and-packet-semantics).

In terms of connection topology, a user would be able to determine a viable channel path from source -> destination using information from the [chain registry](https://github.com/cosmos/chain-registry). They can also independently verify this information via network queries.

### Multihop Relaying

Relayers would deliver channel handshake and IBC packets as they currently do except that they are required to provide proof of the channel path. Relayers would scan packet events for the connectionHops field and determine if the packet is multi-hop by checking the number of hops in the field. If the number of hops is greater than one then the packet is a multi-hop packet and will need extra proof data.

For each multi-hop channel (detailed proof logic below):

1. Scan source chain for IBC messages to relay.
2. Read the connectionHops field in from the scanned message to determine the channel path.

> 3. Lookup connection endpoints via chain registry configuration and ensure each connection in the channel path is updated to include the key/value to be proven on the source chain.

Reviewer 3: Ah yes, I meant to say that the clients associated with the connections in the channel path should be updated to reflect the latest consensus state on the sending chain.

// QueryProofAtHeight returns a proof of the key at the given height and returns the height at which the proof will succeed.
func (TestChain*) QueryProofAtHeight(key string, height int64) (proof []byte, height exported.Height)

// ProofData is a generic proof struct.
type ProofData struct {
    Key   *MerklePath
    Value []]byte
    Proof []byte
}

// MultihopProof defines set of proofs to verify a multihop message.
// Consensus and Connection proofs are ordered from source to destination but not including the
// chain[N-1] consensus state on chainN (dest) since it is already known on the destination chain.
type MultihopProof struct {
    KeyProofIndex uint32           // the index of the consensus state to prove the key/value on, normally 0
    KeyProof *ProofData            // the key/value proof on the source chain (consensusState[0])
    ConsensusProofs []*ProofData   // array of consensus proofs starting with proof of consensusState[0] on chain1
    ConnectionProofs []*ProofData  // array of connection proofs starting with proof of conn[1,0] on chain1 (consensusState[1])
}

// GenerateMultihopProof generates proof of a key/value at the proofHeight on indexed chain (chain0).

> // Chains are provided in order from the sending (source) chain to the verifying chain.

Reviewer 3: updated indexes to reflect chain index reordering.

    // key path to frozen client state
    key := host.FullClientStatePath(counterpartyConnectionEnd.ClientId)
    value := abortTransactionUnless(Marshal(chain1.GetClientState(counterpartyConnectionEnd.ClientId)))

    // query the key/value proof on the next hop chain at the proof height
    keyProof, _ := chain1.QueryProofAtHeight(key, proofHeight.GetRevisionHeight())

    // assign the key/value proof
    proof.KeyProof = &ProofData{
        Key:   nil,    // key to prove constructed during verification using the provided connection state proofs
        Value: value,  // provide the client state value
        Proof: keyProof,
    }

    return &proof
}

// GenerateConsensusProofs generates consensus state proofs starting from the source chain to the N-1'th chain.
// Compute proof for each chain 3-tuple starting from the source chain, C0 to the destination chain, CN
// Step 1: |C0 ---> C1 ---> C2| ---> C3 ... ---> CN
//         (i)    (i+1)    (i+2)
// Step 2: C0 ---> |C1 ---> C2 ---> C3| ... ---> CN
//                 (i)    (i+1)    (i+2)
// Step N-3:  C0 ---> C1 ...  ---> |CN-2 --> CN-1 ---> CN|
//                                   (i)    (i+1)    (i+2)

> // Note: It assumes that clients on  `C1,.., Cn` have been already updated in order.

Reviewer 2: I think this algo will not work if some clients on path are updated in random order after this but before proof collection.

// Relayers may use slightly different algorithms, for example performing the client updates while
// collecting the proofs.
func GenerateConsensusProofs(chains []*Chain) []*ProofData {
    assert(len(chains) > 2)

    var proofs []*ProofData

    // iterate all but the last two chains
    for i := 0; i < len(chains)-2; i++ {

        previousChain := chains[i]   // previous chains state root is on currentChain and is the source chain for i==0.
        currentChain  := chains[i+1] // currentChain is where the proof is queried and generated
        nextChain     := chains[i+2] // nextChain holds the state root of the currentChain

        // height of previous chain client state on current chain
        consensusHeight := currentChain.GetClientStateHeight(previousChain.ClientID)

        // height of current chain state on next chain
        proofHeight := nextChain.GetClientStateHeight(currentChain.ClientID)

>         abortTransactionUnless(consensusHeight >= proofHeight) // ensure that nextChain's client state is update to date

Reviewer 2: ?
```suggestion
        abortTransactionUnless(consensusHeight <= proofHeight) // ensure that nextChain's client state is update to date
```

//                                   (i)    (i+1)    (i+2)
// Note: This implementation assumes the test chains know the connection ids of the previous chain. They are constructed
// such that chain1 knows the ConnectionID to connect to chain0 and so on. The connection end counterparty connection ids
// correspond to the connectionHops field in the IBC channel.
func GenerateConnectionProofs(chains []*Chain) []*ProofData {
    assert(len(chains) > 2)

    var proofs []*ProofData

    // iterate all but the last two chains
    for i := 0; i < len(chains)-2; i++ {

        previousChain := chains[i]  // previous chains state root is on currentChain and is the source chain for i==0.
        currentChain := chains[i+1] // currentChain is where the proof is queried and generated
        nextChain := chains[i+2]    // nextChain holds the state root of the currentChain

        // height of previous chain state on current chain
        connectionHeight := currentChain.GetClientStateHeight(previousChain.ClientID)

        // height of current chain state on next chain
        proofHeight := nextChain.GetClientStateHeight(currentChain.ClientID)

>         abortTransactionUnless(connectionHeight >= proofHeight) // ensure that nextChain's client state is update to date

Reviewer 2: ?
```suggestion
        abortTransactionUnless(connectionHeight <= proofHeight) // ensure that nextChain's client state is update to date
```

7. Repeat step 5, proving `ConsensusState[i]`, and `Conn[i,i+1]` where `i` is the proof index starting with the consensus state on `Chain[N-2]`. `ConsensusState[N-1]` is already known on `Chain[N]`. Note that chains are indexed from sender to receiver and proofs are indexed in the opposite direction to match the connectionHops ordering.
   - Verify ParseClientID(ConsensusProofs[i].Key) == ConnectionEnd.ClientID
   - ConsensusProofs[i].Proof.VerifyMembership(ConsensusState.GetRoot(), ConsensusProofs[i].Key, ConsensusProofs[i].Value)
   - ConnectionProofs[i].Proof.VerifyMembership(ConsensusState.GetRoot(), ConnectionProofs[i].Key, ConnectionProofs[i].Value)
   - ConsensusState = ConsensusProofs[i].Value
   - i++
8. Finally, prove the expected channel or packet commitment in `ConsensusState[N-2]` (sending chain consensus state) on `Chain[1]`

For more details see [ICS4](https://github.com/cosmos/ibc/tree/main/spec/core/ics-004-channel-and-packet-semantics).

### Multi-hop Proof Verification Pseudo Code

Pseudocode proof generation for a channel between `N` chains `C[0] --> C[i] --> C[N]`

```go
// VerifyMultihopMembership verifies a multihop membership proof.

> // Inputs: consensusState - The consensusState for chain[N-1], which is known on the destination chain (chain[N]).

Reviewer 3: same comment as above

//         connectionHops - The expected connectionHops for the channel from the source chain to the destination chain.
//         proof          - The serialized multihop proof data.
//         prefix         - Merkleprefix to be combined with key to generate Merklepath for the key/value proof verification.
//         key            - The key to prove in the indexed consensus state.
//         value          - The value to prove in the indexed consensus state.
func VerifyMultihopMembership(
    consensusState exported.ConsensusState,
    connectionHops []string,
    proof []byte,
    prefix exported.Prefix,
    key string,
    value []byte,
) {
    // deserialize proof bytes into multihop proofs
    proofs := abortTransactionUnless(Unmarshal(proof))
    abortTransactionUnless(len(proofs.ConsensusProofs) >= 1)
    abortTransactionUnless(len(proofs.ConnectionProofs) == len(proofs.ConsensusProofs))

    // verify connection hop ordering and connections are in OPEN state
    abortTransactionUnless(VerifyConnectionHops(proofs.ConnectionProofs, connectionHops))

    // verify intermediate consensus and connection states from destination --> source

>     abortTransactionUnless(VerifyConsensusAndConnectionStates(consensusState, proofs.ConsensusProofs, proofs.ConnectionProofs))

Reviewer 3: In the actual implementation we should probably do this. I left them separate here for clarity, but am happy to combine the logic into the VerifyConsensusAndConnectionStates() if you prefer. lmk.

    // verify a key/value proof on source chain's consensus state.
    abortTransactionUnless(VerifyKeyMembership(proofs, prefix, key, value))
}

// VerifyMultihopNonMembership verifies a multihop non-membership proof.

> // Inputs: consensusState - The consensusState for chain[N-1], which is known on the destination chain (chain[N]).

Reviewer 3: This would imply a reversal of the definition of chain ordering (currently 0 == Sender, N == Receiver). Proof ordering was changed per a previous feedback so that they match the channel hops, but I'm not sure we want to change the chain ordering (side effects would be to rewrite all comments on proof logic and recreate the proof diagrams).

👍🏼 on "destination" --> "receiving"

    connectionHops []string,
    proof []byte,
    prefix exported.Prefix,
    key string,
) {
    // deserialize proof bytes into multihop proofs
    proofs := abortTransactionUnless(Unmarshal(proof))
    abortTransactionUnless(len(proofs.ConsensusProofs) >= 1)
    abortTransactionUnless(len(proofs.ConnectionProofs) == len(proofs.ConsensusProofs))

    // verify connection hop ordering and connections are in OPEN state
    abortTransactionUnless(VerifyConnectionHops(proofs.ConnectionProofs, connectionHops))

    // verify intermediate consensus and connection states from destination --> source
    abortTransactionUnless(VerifyConsensusAndConnectionStates(consensusState, proofs.ConsensusProofs, proofs.ConnectionProofs))

    // verify a key/value proof on source chain's consensus state.
    abortTransactionUnless(VerifyKeyNonMembership(proofs, prefix, key))
}

// VerifyDelayPeriodPassed will ensure that at least delayTimePeriod amount of time and delayBlockPeriod number of blocks have passed
// since consensus state was submitted before allowing verification to continue.
func VerifyDelayPeriodPassed(
    proofHeight exported.Height,
    timeDelay uint64,

>     expectedTimePerBlock uint64,
> ) error {
>     // get time and block delays
>     blockDelay := getBlockDelay(timeDelay, expectedTimePerBlock)

Reviewer 2: ```suggestion
    blockDelay uint64,
) error {

```

Reviewer 3: after simplifying this function isn't even needed 👍🏼

4. Iterate through connection state proofs and verify each connectionEnd is in the OPEN state and check that the connection ids match the channel connectionHops.
5. Verify the intermediate state proofs. Starting with known `ConsensusState[N-1]` at the given `proofHeight` on `ChainN` prove the prior chain's consensus and connection state.
6. Verify that the client id in each consensus state proof key matches the client id in the ConnectionEnd in the previous connection state proof.
7. Repeat step 3, proving `ConsensusState[i-1]`, and `Conn[i-1,i]` until `ConsensusState[0]` on the source chain is proven.
   - Start with i = N-1
   - ConsensusState <- ConsensusProof[i-1].Value on Chain[i]
   - ConnectionEnd <- ConnectionProof[i-1].Value on Chain[i]
   - Verify ParseClientID(ConsensusProofs[i].Key) == ConnectionEnd.ClientID
   - ConsensusProofs[i].Proof.VerifyMembership(ConsensusState.GetRoot(), ConsensusProofs[i].Key, ConsensusProofs[i].Value)
   - ConnectionProofs[i].Proof.VerifyMembership(ConsensusState.GetRoot(), ConnectionProofs[i].Key, ConnectionProofs[i].Value)
   - Set ConsensusState from unmarshalled ConsensusProofs[i].Value

> 8. Finally, prove the expected channel or packet commitment in `ConsensusState[0]`

Reviewer 2: These comments should be updated to reflect that first proof is for the receiving chain and last for the source (or misbehaving) chain.

        abortTransactionUnless(connectionID == connectionHops[i+1])

        // 2. check that the connectionEnd's are in the OPEN state.
        abortTransactionUnless(connectionEnd.GetState() == int32(connectiontypes.OPEN))
    }
}

// VerifyConsensusAndConnectionStates verifies the state of each intermediate consensus and connection state
// starting from the receiving chain and finally proving the sending chain consensus and connection state.
func VerifyConsensusAndConnectionStates(
    consensusState exported.ConsensusState,
    consensusProofs []*ProofData,
    connectionProofs []*ProofData,
) {
    // iterate through proofs to prove from executing chain (receiver) to counterparty chain (sender)
    var clientID string
    for i := 0; i < len(consensusProofs); i++ {

        // check the clientID from the previous connection is part of the next
        // consensus state's key path.
        abortTransactionUnless(VerifyClientID(clientID, consensusProofs[i].Key))

        // prove the consensus state of chain[i] on chain[i-1]
        consensusProof := abortTransactionUnless(Unmarshal(consensusProofs[i].Proof))
        abortTransactionUnless(consensusProof.VerifyMembership(

>             commitmenttypes.GetSDKSpecs(),

Reviewer 4: One question:
Does this assume that all proof specs for membership verification for each chain are equal? If not, then I think we need to refer to the client state as well as the consensus state for each chain.

Reviewer 3: Yeah currently it is assuming proof specs are equal.

Reviewer 2: imo this is a very important point and we should make it very clear in the spec.

---
ics: 33
title: Multi-hop Channel
stage: draft
required-by: 4
category: IBC/TAO
kind: instantiation
author: Derek Shiell <derek@polymerlabs.org>, Wenwei Xiong <wenwei@polymerlabs.org>, Bo Du <bo@polymerlabs.org>
created: 2022-11-11
modified: 2022-12-16
---

## Synopsis

> This document describes a standard for multi-hop IBC channels. Multi-hop channels specifies a way to route messages across a path of IBC enabled blockchains utilizing multiple pre-existing IBC connections.

Reviewer 2: ```suggestion
This document describes a standard for multi-hop IBC channels. Multi-hop channels specify a way to route messages across a path of IBC enabled blockchains utilizing multiple pre-existing IBC connections.
```

### Definitions

Associated definitions are as defined in referenced prior standards (where the functions are defined), where appropriate.

`Connection` is as defined in [ICS 3](https://github.com/cosmos/ibc/tree/main/spec/core/ics-003-connection-semantics).

`Channel` is as defined in [ICS 4](https://github.com/cosmos/ibc/tree/main/spec/core/ics-004-channel-and-packet-semantics).

`Channel Path` is defined as the path of connection IDs along which a channel is defined.

`Connection Hop` is defined as the connection ID of the connection between two chains along a channel path.

### Desired Properties

- IBC channel handshake and message packets should be able to be utilize pre-existing connections to form a logical proof chain to relay messages between unconnected chains.

> - Relaying for a multi connection IBC channel should NOT require additional writes to intermediate hops.

Reviewer 2: Not sure about this as we do need client updates. 
```suggestion
- Relaying for a multi connection IBC channel should require minimal number of additional writes to state on intermediate hops.
```

Reviewer 3: True, I think this was a statement to contrast with prior multi-hop designs which would have written ICS-04 packet data to intermediate hops. I'll update the wording to capture all the nuance.

- Minimal additional required state and changes to core and app IBC specs.
- Retain desired properties of connection, channel and packet definitions.
- Retain backwards compatibility for messaging over a single connection hop.

## Technical Specification

The bulk of the spec will be around proof generation and verification. IBC connections remain unchanged. Additionally, channel handshake and packet message types as well as general round trip messaging semantics and flow will remain the same. There is additional work on the verifier side on the receiving chain as well as the relayers who need to query for proofs.

> Messages passed over multiple hops require proof of the connection path from source chain to receiving chain as well as the packet commitment on the source chain. The connection path is proven by verifying the connection state and consensus state of each connection in the path to the receiving chain. On a high level, this can be thought of as a channel path proof chain where the receiving chain can prove a key/value on the source chain by iteratively proving each connection and consensus state in the channel path starting with the consensus state associated with the final client on the receiving chain. Each subsequent consensus state and connection is proven until the source chain's consensus state is proven which can then be used to prove the desired key/value on the source chain.

Reviewer 2: ```suggestion
Messages passed over multiple hops require proof of the connection path from source chain to receiving chain as well as the packet commitment on the source chain. The connection path is proven by verifying the connection state and consensus state of each connection in the path to the receiving chain. On a high level, this can be thought of as a channel path chained proof where the receiving chain can prove a key/value on the source chain by iteratively proving each connection and consensus state in the channel path starting with the consensus state associated with the final client on the receiving chain. Each subsequent consensus state and connection is proven until the source chain's consensus state is proven which can then be used to prove the desired key/value on the source chain.
```

### Channel Handshake and Packet Messages

> For both channel handshake and packet messages, additional connection hops are defined in the pre-existing `connectionHops` field. The connection IDs along the channel path must be pre-existing and in the `OPEN` state to guarantee delivery to the correct recipient. See `Path Forgery Protection` for more info.

Reviewer 2: ```suggestion
For both channel handshake and packet messages, additional connection hops are defined in the pre-existing `connectionHops` field. The connections along the channel path must exists and in the `OPEN` state to guarantee delivery to the correct recipient. See `Path Forgery Protection` for more information.
```

> The spec for channel handshakes and packets remains the same. See [ICS 4](https://github.com/cosmos/ibc/tree/main/spec/core/ics-004-channel-and-packet-semantics).

Reviewer 2: There are changes to ics-04. So maybe rephrase?

In terms of connection topology, a user would be able to determine a viable channel path from sender -> receiver using information from the [chain registry](https://github.com/cosmos/chain-registry). They can also independently verify this information via network queries.

### Multihop Relaying

Relayers would deliver channel handshake and IBC packets as they currently do except that they are required to provide proof of the channel path. Relayers would scan packet events for the connectionHops field and determine if the packet is multi-hop by checking the number of hops in the field. If the number of hops is greater than one then the packet is a multi-hop packet and will need extra proof data.

For each multi-hop channel (detailed proof logic below):

1. Scan source chain for IBC messages to relay.
2. Read the connectionHops field in from the scanned message to determine the channel path.
3. Lookup connection endpoints via chain registry configuration and update the clients associated with the connections in the channel path to reflect the latest consensus state on the sending chain including the key/value to be proven.

> 4. Query for proof of connection, and consensus state for each intermediate connection in the channel path.

Reviewer 2: these are done in one pass, no? Maybe this needs to be updated as per pseudocode?

Reviewer 3: I will update this section. The algorithm is now 2 pass.

1. query each chain in the channel path to determine proof heights and update clients as required
2. query for proofs at the determined proof heights

> 5. Query proof of packet commitment or handshake message commitment on source chain.

Reviewer 2: ```suggestion
5. Query proof of packet or handshake message commitment on source chain.
```

6. Submit proofs and data to RPC endpoint on receiving chain.

Relayers are connection topology aware with configurations sourced from the [chain registry](https://github.com/cosmos/chain-registry).

### Proof Generation & Verification

![proof_generation.png](proof_generation.png)
Graphical depiction of proof generation.

![relayer_calc_update_heights.png](relayer_calc_update_heights.png)
During the first phase of querying a multi-hop proof the relayer searches for the minimum consensus height that can prove the previous chain state and is also provable by the next chain in the channel path.

![relayer_query_proof_and_submit.png](relayer_query_proof_and_submit.png)
In the second phase of querying a multi-hop proof, the relayer queries proofs of the client and connection states as well as the key/value on the source chain at the heights determined in the first phase.

![proof_query_algorithm.png](proof_query_algorithm.png)
Multi-hop proof query algorithm.

![proof_verification.png](proof_verification.png)

> Multi-hop proof verfication logic. Pseudocode proof generation for a channel between `N` chains `C[0] --> C[i] --> C[N]`

Reviewer 2: ```suggestion
Multi-hop proof verfication logic. 

Proof generation pseudocode for a channel with a channel path with `N` chains, `C[0] --> C[i] --> C[N]`
```

5. Verify the intermediate state proofs. Starting with known `ConsensusState[0]` at the given `proofHeight` on `Chain[1]` prove the prior chain's consensus and connection state.
6. Verify that the client id in each consensus state proof key matches the client id in the ConnectionEnd in the previous connection state proof.
7. Repeat step 5, proving `ConsensusState[i]`, and `Conn[i,i-1]` where `i` is the proof index starting with the consensus state on `Chain[2]`. `ConsensusState[1]` is already known on `Chain[0]`. Note that chains are indexed from executing (verifying) chain to  and proofs are indexed in the opposite direction to match the connectionHops ordering.
   - Verify ParseClientID(ConsensusProofs[i].Key) == ConnectionEnd.ClientID
   - ConsensusProofs[i].Proof.VerifyMembership(ConsensusState.GetRoot(), ConsensusProofs[i].Key, ConsensusProofs[i].Value)
   - ConnectionProofs[i].Proof.VerifyMembership(ConsensusState.GetRoot(), ConnectionProofs[i].Key, ConnectionProofs[i].Value)
   - ConsensusState = ConsensusProofs[i].Value

>    - i++

Reviewer 2: these could be done in one pass, no?

Reviewer 2: github missed the line selection, i am referring to 4 and 5-7.

Reviewer 2: on the other hand, thinking about it more, maybe it is better to have it this way as we catch in a cheap way bad connection on last hops without first verifying proofs for nothing.

Reviewer 3: yeah, step 4 iterates over only the connection state proofs (without verifying). 5-7 iterates over consensus states and uses the consensus states to recursively verify consensus and connection proofs.

// Inputs: consensusState - The consensusState for chain[1], which is known on the receiving chain (chain[0]).
//         connectionHops - The expected connectionHops for the channel from the receiving chain to the sending chain.
//         proof          - The serialized multihop proof data.
//         prefix         - Merkleprefix to be combined with key to generate Merklepath for the key/value proof verification.
//         key            - The key to prove absent in the indexed consensus state
func VerifyMultihopNonMembership(
    consensusState exported.ConsensusState,
    connectionHops []string,
    proof MultihopProof,
    prefix exported.Prefix,
    key string,
) {
    // deserialize proof bytes into multihop proofs
    proofs := abortTransactionUnless(Unmarshal(proof))
    abortTransactionUnless(len(proofs.ConsensusProofs) >= 1)
    abortTransactionUnless(len(proofs.ConnectionProofs) == len(proofs.ConsensusProofs))

    // verify connection hop ordering and connections are in OPEN state
    abortTransactionUnless(VerifyConnectionHops(proofs.ConnectionProofs, connectionHops))

    // verify intermediate consensus and connection states from receiver --> sender

>     abortTransactionUnless(VerifyIntermediateStateProofs(consensusState, proofs.ConsensusProofs, proofs.ConnectionProofs))

Reviewer 2: related to the previous comment we could have VerifyMultiHopProofs or something like that, would verify both the connection proof and consensus proof in in one step of the iterator.
Can even change MultihopProof to be a slice of (connection_proof, cosnensus_proof) elements.
Do you have a pointer to the .protos for these in your code?

Reviewer 3: Here is the proto def: https://github.com/polymerdao/ibc-go/blob/polymer/multihop-v7/proto/ibc/core/channel/v1/tx.proto#L283

In the spec, it becomes harder to understand if the state and verification logic are combined so I left them separate. The current implementation also leaves them separate to do the cheaper connection state check first and then verify the proofs. However, optimizations could potentially be made to combine these if desired.

    // verify a key/value proof on source chain's consensus state.

>     abortTransactionUnless(VerifyKeyNonMembership(consensusState, proofs, prefix, key))

Reviewer 2: related to the previous comment we could have `VerifyMultiHopProofs` or something like that, would verify both the connection proof and consensus proof in in one step of the iterator.
Can even change `MultihopProof` to be a slice of (connection_proof, cosnensus_proof) elements. 
Do you have a pointer to the .protos for these in your code?
